src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -751,3 +886,307 @@ RouterExecutorEnd(QueryDesc *queryDesc)
 	queryDesc->estate = NULL;
 	queryDesc->totaltime = NULL;
 }
+
+
+/*
+ * InstallRouterExecutorShmemHook simply installs a hook (intended to be called
+ * once during backend startup), which will itself register all the transaction
+ * callbacks needed by this executor.
+ */
+void
+InstallRouterExecutorShmemHook(void)
+{
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = RegisterRouterExecutorXactCallbacks;
+}
+
+
+/*
+ * RegisterRouterExecutorXactCallbacks registers (sub-)transaction callbacks
+ * needed by this executor before calling any previous shmem startup hooks.
+ */
+static void
+RegisterRouterExecutorXactCallbacks(void)
+{
+	RegisterXactCallback(RouterTransactionCallback, NULL);
+	RegisterSubXactCallback(RouterSubtransactionCallback, NULL);
+
+	if (prev_shmem_startup_hook != NULL)
+	{
+		prev_shmem_startup_hook();
+	}
+}
+
+
+/*
+ * RouterTransactionCallback handles committing or aborting remote transactions
+ * after the local one has committed or aborted. It only sends COMMIT or ABORT
+ * commands to still-healthy remotes; the failed ones are marked as inactive if
+ * after a successful COMMIT (no need to mark on ABORTs).
+ */
+static void
+RouterTransactionCallback(XactEvent event, void *arg)
+{
+	if (xactParticipantHash == NULL)
+	{
+		return;
+	}
+
+	switch (event)
+	{
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_COMMIT:
+#endif
+		case XACT_EVENT_COMMIT:
+		{
+			bool commit = true;
+
+			ExecuteTransactionEnd(commit);
+
+			break;
+		}
+
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_ABORT:
+#endif
+		case XACT_EVENT_ABORT:
+		{
+			bool commit = false;
+
+			ExecuteTransactionEnd(commit);
+
+			break;
+		}
+
+		/* no support for prepare with long-running transactions */
+		case XACT_EVENT_PREPARE:
+		case XACT_EVENT_PRE_PREPARE:
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot prepare a transaction that modified "
+								   "distributed tables")));
+
+			break;
+		}
+
+#if (PG_VERSION_NUM >= 90500)
+		case XACT_EVENT_PARALLEL_PRE_COMMIT:
+#endif
+		case XACT_EVENT_PRE_COMMIT:
+		{
+			if (subXactAbortAttempted)
+			{
+				subXactAbortAttempted = false;
+
+				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								errmsg("cannot ROLLBACK TO SAVEPOINT in transactions "
+									   "which modify distributed tables")));
+			}
+
+			/* leave early to avoid resetting transaction state */
+			return;
+		}
+	}
+
+	/* reset transaction state */
+	hash_destroy(xactParticipantHash);
```

Comment:
BTW, I couldn't easily find a `hash_clear` thing that would erase hash contents without erasing the hash itself. Does such a thing exist? If so, I could keep the hash for the entire session (rather than recreating it for each top-level transaction).